resharper/resharper-unity/src/CSharp/Feature/Services/QuickFixes/MoveQuickFixes/MonoBehaviourMoveUtil.cs

```diff
@@ -0,0 +1,263 @@
+                var declaration = (IMethodDeclaration)factory.CreateTypeMemberDeclaration("void $0(){}", methodName);
+
+                result = classDeclaration.AddClassMemberDeclarationAfter(declaration, classDeclaration.FieldDeclarations.FirstOrDefault());
+               
+            }
+            return result;
+        }
+
+        public static bool IsExpressionAccessibleInScript([NotNull]ICSharpExpression expression)
+        {
+            expression.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            Shell.Instance.GetComponent<IShellLocks>().AssertMainThread();
+            
+            if (!expression.IsValid())
+                return false;
+            
+            var methodDeclaration = expression.GetContainingNode<IMethodDeclaration>();
+            if (methodDeclaration == null)
+                return false;
+            
+            var statement = expression.GetContainingStatementLike();
+            if (statement == null)
+                return false;
+
+            var classDeclaration = expression.GetContainingNode<IClassDeclaration>();
+            if (classDeclaration == null)
+                return false;
+
+            return IsAvailableToMoveFromMethodToMethod(expression);
+        }
+
+        /// <summary>
+        /// Is node available to move outside the loop
+        /// </summary>
+        public static bool IsAvailableToMoveFromLoop([NotNull] ITreeNode toMove, [NotNull] ILoopStatement loop)
+        {
+            toMove.GetPsiServices().Locks.AssertReadAccessAllowed();
+
+            var sourceFile = toMove.GetSourceFile();
+            if (sourceFile == null)
+                return false;
+            
+            var loopStartOffset = loop.GetTreeStartOffset().Offset;
+            return IsAvailableToMoveInner(toMove, declaredElement => 
+                    declaredElement.GetDeclarationsIn(sourceFile)
+                    .FirstOrDefault(declaration => declaration.GetSourceFile() == sourceFile)?.GetTreeStartOffset().Offset < loopStartOffset
+            );
+        }
+
+        /// <summary>
+        /// Is node available to move from one method to another in MonoBehaviour class 
+        /// </summary>
+        public static bool IsAvailableToMoveFromMethodToMethod([NotNull] ITreeNode toMove)
+        {
+            toMove.GetPsiServices().Locks.AssertReadAccessAllowed();
+            
+            return IsAvailableToMoveInner(toMove);
+        }
+        
+        private static bool IsAvailableToMoveInner([NotNull] ITreeNode toMove, [CanBeNull] Func<IDeclaredElement, bool> isElementIgnored = null)
+        {
+            var nodeEnumerator = toMove.ThisAndDescendants();
+            while (nodeEnumerator.MoveNext())
+            {
+                var current = nodeEnumerator.Current;
+                IDeclaredElement declaredElement;
+                switch (current)
+                {
+                    case IDeclaration declaration:
```

Comment:
It looks like the current code will disallow the following based on the local variable declaration presence:
`M(() => { var x = 0; });`
This doesn't look intended, I guess it shouldn't process declarations in nested scopes. 

If this declaration check should disallow code like this: `M(out var x)` there's an existing extension method `ICSharpExpression.GetDeclaredExpressionVariables` to collect such variables which does not traverse closures.